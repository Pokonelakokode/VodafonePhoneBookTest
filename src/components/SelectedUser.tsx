import * as React from "react";
import {useSelector} from "react-redux";
import {mainSelector} from "../actions";

interface IProps {
    selectedUser: number
}

const SelectedUser: React.FC<IProps> = ({selectedUser}) => {
    const {users} = useSelector(mainSelector);
    const user = users.find(user => user.id === selectedUser);
    if(!user) return <h1>USER NOT FOUND</h1>;
    return (
        <div className="row">
            <h1>{`${user.name}'s Profile`}</h1>

        </div>
    )
};

export default SelectedUser