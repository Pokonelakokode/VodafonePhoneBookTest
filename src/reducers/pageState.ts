import {IPageStateActions, PageStateActionTypes} from "../actions/pageState";
import {getHash} from "../utils";

export type PageState = {
    selectedUser: number | null,
    search: string,
    letterFilter: string
}

const initialState:PageState = {
    selectedUser: parseInt(getHash().user) || null,
    search: "",
    letterFilter: ""
};


export default function pageState(state = initialState, action: IPageStateActions) {
    switch (action.type) {
        case PageStateActionTypes.SET:
            return {...state,[action.key]:action.data};
        default:
            return state
    }
}

