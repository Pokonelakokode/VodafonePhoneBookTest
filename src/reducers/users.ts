import {IUserActions, UserActionTypes} from "../actions/users";
import { User } from "../types";
import {loadFromStorage} from "../utils";

export default function users(state = loadFromStorage(), action: IUserActions):User[] {
    switch (action.type) {
        case UserActionTypes.LOAD:
            return action.users;
        case UserActionTypes.SET:
            return state.map(user => user.id === action.user.id ? action.user : user);
        case UserActionTypes.ADD:
            return [...state,action.user];
        case UserActionTypes.DELETE:
            return state.filter(user => user.id !== action.id);
        default:
            return state
    }
}