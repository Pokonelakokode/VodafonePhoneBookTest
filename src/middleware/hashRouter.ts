import {Dispatch, Middleware} from "redux";
import {RootActions} from "../actions";
import {PageStateActionTypes} from "../actions/pageState";
import {setHash} from "../utils";

export const hashRouter:Middleware = () => (next:Dispatch) => (action:RootActions) => {
    next(action);
    if(action.type === PageStateActionTypes.SET && action.key === "selectedUser"){
        action.data ? setHash({user:action.data}) : setHash();
    }
};