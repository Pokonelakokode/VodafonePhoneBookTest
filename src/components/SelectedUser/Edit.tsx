import * as React from "react";
import {useEffect, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {mainSelector} from "../../actions";
import {addUser, deleteUser, getUserOrNew, saveUser} from "../../actions/users";
import {bindActionCreators} from "redux";
import {selectUser} from "../../actions/pageState";
import {validateEmail} from "../../utils";

interface IProps {
    selectedUser: number
}


const Edit: React.FC<IProps> = ({selectedUser}) => {
    if(selectedUser === null) return <h1>No such user</h1>;
    const {users} = useSelector(mainSelector);
    const actions = bindActionCreators({saveUser, addUser, selectUser, deleteUser}, useDispatch());

    const [user, setUser] = useState(getUserOrNew(users, selectedUser)[0]);
    const [edit, setEdit] = useState(!user.id);
    useEffect(() => {
        let [user, isNew] = getUserOrNew(users, selectedUser);
        setUser(user);
        setEdit(isNew);
    }, [selectedUser]);
    const editCancel = () => {
        if (edit) setUser(getUserOrNew(users, selectedUser)[0]);
        setEdit(!edit);
    };
    const _deleteUser = () => {
        if (confirm(`Are you sure you want to delete ${user.name} from contacts?`)) {
            actions.deleteUser(user.id);
            actions.selectUser(null)
        }
    };
    const nameValid = user.name.length > 0;
    const emailValid = validateEmail(user.email);
    const phoneValid = user.phone.length >= 7;
    const _saveUser = () => {
        let isNew: boolean;
        if(!nameValid || !emailValid || !phoneValid) return alert("Some of your data is invalid");
        [user.id, isNew] = !user.id ? [users.reduce((max: number, _user) => _user.id > max ? _user.id : max, -1) + 1, true] : [user.id, false];
        if (isNew) {
            actions.addUser(user);
            actions.selectUser(user.id);
        } else {
            actions.saveUser(user);
        }
        setEdit(false);
    };

    return (
        <div>
            <h1 className="text-center">{!user.id ? "Add New Contact" : `${edit ? "Edit " : ""}${user.name}'s Profile`}</h1>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Name</label>
                <div className="col-sm-10">
                    <input value={user.name}
                           placeholder="Enter Name"
                           readOnly={!edit} type="text"
                           className={`${nameValid ? "" : "is-invalid "}form-control`}
                           required
                           onChange={(e) => setUser({...user, name: e.target.value})}
                    />
                    {!nameValid && <div className="invalid-feedback d-block">Name is too short</div>}
                </div>

            </div>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Email</label>
                <div className="col-sm-10">
                    <input value={user.email}
                           placeholder="Enter Email"
                           readOnly={!edit} type="email"
                           className={`${emailValid ? "" : "is-invalid "}form-control`}
                           onChange={(e) => setUser({...user, email: e.target.value})}
                    />
                    {!emailValid && <div className="invalid-feedback d-block">Email address is invalid</div>}
                </div>
            </div>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Phone</label>
                <div className="col-sm-10">
                    <input value={user.phone} placeholder="Enter Phone"
                           readOnly={!edit} type="text"
                           required
                           className={`${phoneValid ? "" : "is-invalid "}form-control`}
                           onChange={(e) => setUser({...user, phone: e.target.value})}
                    />
                    {!phoneValid && <div className="invalid-feedback d-block">Phone number have to be at least 7 digits</div>}
                </div>
            </div>
            <div className="row justify-content-around">
                {user.id &&
                <button onClick={editCancel} className="btn btn-primary">{edit ? "Cancel" : "Edit"}</button>}
                <button onClick={() => actions.selectUser(null)} className="btn btn-primary">Back</button>
                {edit ?
                    <button onClick={_saveUser} className="btn btn-danger">{user.id ? "Save" : "Create"}</button>
                    :
                    user.id && <button onClick={_deleteUser} className="btn btn-danger">Delete</button>
                }
            </div>

        </div>
    )
};

export default Edit