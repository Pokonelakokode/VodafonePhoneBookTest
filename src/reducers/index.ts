import { combineReducers } from "redux";
import pageState from "./pageState";
import users from "./users";

export const rootReducer = combineReducers({
    pageState,
    users
});

export type RootState = ReturnType<typeof rootReducer>