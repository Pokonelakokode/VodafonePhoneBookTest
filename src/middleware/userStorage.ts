import {Dispatch, Middleware, MiddlewareAPI} from "redux";
import {RootActions} from "../actions";
import {UserActionTypes} from "../actions/users";
import {saveToStorage} from "../utils";
import {RootState} from "../reducers";

export const userStorage:Middleware = (store: MiddlewareAPI<any,RootState>) => (next:Dispatch) => (action:RootActions) => {
    const {DELETE, ADD, SET} = UserActionTypes;
    next(action);
    if(action.type === DELETE || ADD || SET){
        saveToStorage(store.getState().users)
    }
};

export default userStorage