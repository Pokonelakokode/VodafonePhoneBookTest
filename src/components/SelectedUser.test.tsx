import * as React from "react";
import * as renderer from "react-test-renderer"
import {createStore, applyMiddleware} from "redux";
import {rootReducer} from "../reducers";
import {hashRouter} from "../middleware/hashRouter";
import userStorage from "../middleware/userStorage";
import { Provider } from "react-redux";
import SelectedUser from "./SelectedUser";

describe("SelectedUser",() => {
    const store = createStore(rootReducer, applyMiddleware(hashRouter, userStorage));
    it('should render snapshot', function () {
        const component = renderer.create(
            <Provider store={store}>
                <SelectedUser selectedUser={1}/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});