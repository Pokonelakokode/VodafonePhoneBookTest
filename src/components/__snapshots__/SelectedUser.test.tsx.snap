// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedUser should render snapshot 1`] = `
<div
  className="row"
>
  <h1>
    Leanne Graham's Profile
  </h1>
</div>
`;
